api/config/v1alpha1/helpers.go

```diff
@@ -248,3 +254,34 @@ func (r *EnvoyGatewayProvider) GetEnvoyGatewayKubeProvider() *EnvoyGatewayKubern
 
 	return r.Kubernetes
 }
+
+// ApplyMergePatch applies a merge patch to a deployment based on the merge type
+func (deployment *KubernetesDeploymentSpec) ApplyMergePatch(old *appv1.Deployment) (*appv1.Deployment, error) {
+	if deployment.Patch == nil {
+		return old, nil
```

Comment:
This returns the same struct pointer if the patch is nil, but a new struct pointer if the patch is applied.
That's probably fine, just pointing it out in case it matters.

Reply:
Good callout - seems to be fine based on the logic surrounding the actual update: https://github.com/gravitational/gateway/blob/e35248374fc1acfe1072d7777edb1995f2f2a184/internal/infrastructure/kubernetes/deployment.go#L33-L42

In upstream where this PRs code comes from they are using a different update mechanism:

https://github.com/Xunzhuo/gateway/blob/29b5eaa6a209adb4e336b3c773dc92bdb647c891/internal/infrastructure/kubernetes/infra_resource.go#L109

https://github.com/envoyproxy/gateway/blob/ac86045216b4c2a432d39ca365664f6aabdf2923/internal/infrastructure/kubernetes/infra_client.go#L27-L31

I am seeing a fair number of these errors after updating the `EnvoyProxy` CR (when running the `envoy-gateway` build based on this PR):
```
failed to create new infra	{"runner": "infrastructure", "error": "failed to update deployment gateway-system/envoy-gateway-system-tenant-gateway-67617465: Operation cannot be fulfilled on deployments.apps \"envoy-gateway-system-tenant-gateway-67617465\": the object has been modified; please apply your changes to the latest version and try again"}
```

I was assuming this was a symptom of having two `envoy-gateway` pods watching the same resources. Will try switching back to the version we're currently running to see if the logs look any different.

Reply:
Should we be ignoring the error here?
https://github.com/gravitational/gateway/pull/63/files#diff-9077c1e217cd07ad79ea322de3b24ee92c36045aa1f26e28a36c8378a0e4474eR102

Reply:
If we return the error, the deployment doesn't get updated at all which is reasonable behaviour. 1405381f

Reply:
https://github.com/gravitational/gateway/pull/63/files#r1873645382

Reply:
Nevermind, I missed one of the two upstream PRs: https://github.com/envoyproxy/gateway/pull/2374

Ignore my suggestions and use what's upstream, since we'll adopt that code eventually anyways.

Reply:
> Will try switching back to the version we're currently running to see if the logs look any different.

Tested with `public.ecr.aws/gravitational/envoy-gateway:v0.4.0-teleport.6-11-g98d3233` and still get the conflict errors on Deployment reconciliation.

Seeing the same error logs in prod: 
![image](https://github.com/user-attachments/assets/caa8009c-79ea-4628-8d12-ee216dd38fa3)